### PR TITLE
Add CircleCI support

### PR DESCRIPTION
**Pull request by ivan4th**
Let's keep Travis for a little while to have a chance of comparing the stability of these CI systems.

---
This change is [Reviewable](https://reviewable.io/reviews/mirantis/virtlet/359)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        name: Ensure prerequisites
+        command: |
+          apt-get -qq update
+          apt-get install -y libxml2-utils jq curl ca-certificates openssh-client rsync socat wget
+
+    - run:
+        name: Install Docker client
+        command: |
+          set -x
+          VER="17.03.0-ce"
```
Comment (holser) [inline .circleci/config.yml]:
```
    environment:
      - DOCKER_VERSION: "17.03.0-ce"
```

to separate code from data.

Comment (ivan4th) [inline .circleci/config.yml]: Frankly I just copied CircleCI's official example verbatim. But yes, we may improve it.

**Comment by holser** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[.circleci/config.yml, line 21 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko81mtmcCYmKrKTlrQI-r1-21:-Ko86foh350qB1VRlMLa:b-zdnbcs) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L21)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

Frankly I just copied CircleCI's official example verbatim. But yes, we may improve it.
</blockquote></details>

I think it's different example. Use example from https://circleci.com/docs/2.0/configuration-reference/

---

*[.circleci/config.yml, line 22 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko86wgyqPjvkxi_okUc:-Ko86wgyqPjvkxi_okUd:bszxple) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L22)):*
> ```YAML
>           set -x
>           VER="17.03.0-ce"
>           curl -L -o /tmp/docker-$VER.tgz https://get.docker.com/builds/Linux/x86_64/docker-$VER.tgz
> ```

Please use -sSL instead of -L
From man:
 -s, --silent
-S, --show-error

---

*[.circleci/config.yml, line 33 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko87Oj-BxR_UIhMFR4a:-Ko87Oj-BxR_UIhMFR4b:b-vcm4s4) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L33)):*
> ```YAML
>           docker run --rm --pid=host -it --privileged ubuntu:xenial \
>             nsenter -t 1 -m -u -i -n /bin/bash -c \
>             "service apparmor stop && service apparmor teardown && update-rc.d -f apparmor remove && apt-get remove -y apparmor"
> ```

Can you split this line to 4 different lines? Thank you in advance
 "service apparmor stop && \
   service apparmor teardown && \
   update-rc.d -f apparmor remove && \
   apt-get remove -y apparmor"

---

*[.circleci/config.yml, line 62 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko87nuokqODRW398qK6:-Ko87nuokqODRW398qK7:b5fmsjg) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L62)):*
> ```YAML
>         command: |
>           build/portforward.sh 8080&
>           NONINTERACTIVE=1 NO_VM_CONSOLE=1 INJECT_LOCAL_IMAGE=1 BASE_LOCATION="$PWD" deploy/demo.sh
> ```

Do we need to move these variables to environments: ? In that case we'll be able to override them without modifying config.yaml

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/359)*

**Comment by holser** [comment]
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/359)*

**Comment by ivan4th** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[.circleci/config.yml, line 21 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko81mtmcCYmKrKTlrQI-r1-21:-Ko89Y28lafHvMZII7DF:b5cf0n5) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L21)):*
<details><summary><i>Previously, holser (Sergii Golovatiuk) wrote…</i></summary><blockquote>

I think it's different example. Use example from https://circleci.com/docs/2.0/configuration-reference/
</blockquote></details>

The example comes from here: https://circleci.com/docs/2.0/building-docker-images/#example

---

*[.circleci/config.yml, line 22 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko86wgyqPjvkxi_okUc:-Ko89a1muKKSMpJSmJsY:bdtxish) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L22)):*
<details><summary><i>Previously, holser (Sergii Golovatiuk) wrote…</i></summary><blockquote>

Please use -sSL instead of -L
From man:
 -s, --silent
-S, --show-error

</blockquote></details>

See above, I usually do `-sSL` too, in this case just copied the example

---

*[.circleci/config.yml, line 33 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko87Oj-BxR_UIhMFR4a:-Ko89fBABhzugdElyRcb:b32wtd) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L33)):*
<details><summary><i>Previously, holser (Sergii Golovatiuk) wrote…</i></summary><blockquote>

Can you split this line to 4 different lines? Thank you in advance
 "service apparmor stop && \
   service apparmor teardown && \
   update-rc.d -f apparmor remove && \
   apt-get remove -y apparmor"
</blockquote></details>

ok

---

*[.circleci/config.yml, line 62 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko87nuokqODRW398qK6:-Ko89kuFDB9r2AkJ37aF:b-ue70i3) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L62)):*
<details><summary><i>Previously, holser (Sergii Golovatiuk) wrote…</i></summary><blockquote>

Do we need to move these variables to environments: ? In that case we'll be able to override them without modifying config.yaml
</blockquote></details>

These vars modify `demo.sh` behavior so as to make it suitable for the CI env. Without them, `demo.sh` wouldn't work here. So they're not really job settings but just means of making `demo.sh` work without real user interaction. Probably the better approach would be some kind of `--ci` flags to demo.sh but that's cosmetic stuff that on the other hand would require some extra testing (that probably includes multiple manual demo.sh invocations, too)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/359)*

**Comment by ivan4th** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[.circleci/config.yml, line 62 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko87nuokqODRW398qK6:-Ko8AUVpm5_U8NmHfeMq:b-415747) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L62)):*
> would be some kind of --ci flags
err --- would be passing some kind of --ci flag

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/359)*

**Comment by jellonek** [comment]
Reviewed 2 of 2 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

*[.circleci/config.yml, line 62 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko87nuokqODRW398qK6:-KoBX373dJFA5Bd1VAeD:b6chstr) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L62)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

> would be some kind of --ci flags
err --- would be passing some kind of --ci flag

</blockquote></details>

Please add followup issue for hiding that in `demo.sh` script under mentioned flag.

---

*[build/portforward.sh, line 17 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/359#-KoBXq-3Koos0dTnScjE:-KoBXq-3Koos0dTnScjF:blimzhu) ([raw file](https://github.com/mirantis/virtlet/blob/f2c7183f406f5e3de9f97d2b27162806f7a1c364/build/portforward.sh#L17)):*
> ```Shell
>         EXEC:"'docker exec -i portforward socat STDIO TCP-CONNECT:localhost:${1}'"
> else
>   echo "Must specify either start or the port number" >&2
> ```

Even if output value it's not used now - this should end with `exit 1` or any other  than `0` value.

---

*[pkg/criproxy/proxy_test.go, line 1160 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/359#-KoBYk36O2QFjs0Yj9Yl:-KoBYk36O2QFjs0Yj9Ym:bt73itg) ([raw file](https://github.com/mirantis/virtlet/blob/f2c7183f406f5e3de9f97d2b27162806f7a1c364/pkg/criproxy/proxy_test.go#L1160)):*
> ```Go
> 
> 	// skip for PRs, run on master only
> 	t.Skip("skipping flaky test -- will only run on master branch")
> ```

This also could be written in form like:
```
func skip() {
    t.Skip("skipping flaky test -- will only run on master branch")
}

switch {
case os.Getenv("CIRCLECI") == "true":
    if os.Getenv("CIRCLE_PULL_REQUEST") != "" || os.Getenv("CIRCLE_BRANCH") != "master" {
        skip()
    }
case osGetenv("TRAVIS") == "true":
    if os.Getenv("TRAVIS_PULL_REQUEST") != "false" || os.Getenv("TRAVIS_BRANCH") != "master" {
        skip()
    }
}
```
With that we have a single point of exit and IMO it's a bit more readable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/359)*

**Comment by ivan4th** [comment]
Review status: 1 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[.circleci/config.yml, line 21 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko81mtmcCYmKrKTlrQI-r1-21:-KoTyZn5B9lG2zsJFlJV:b-6trd2e) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L21)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

The example comes from here: https://circleci.com/docs/2.0/building-docker-images/#example
</blockquote></details>

Fixed version ref

---

*[.circleci/config.yml, line 22 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko86wgyqPjvkxi_okUc:-KoTybv5e8-VBfxdnWxQ:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L22)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

See above, I usually do `-sSL` too, in this case just copied the example
</blockquote></details>

Done.

---

*[.circleci/config.yml, line 33 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko87Oj-BxR_UIhMFR4a:-KoTyLwLf00eVRk_PyPD:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L33)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

ok
</blockquote></details>

Done.

---

*[.circleci/config.yml, line 62 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko87nuokqODRW398qK6:-KoTyCHCXN-tn_h1vP1k:b-uobjq0) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L62)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Please add followup issue for hiding that in `demo.sh` script under mentioned flag.
</blockquote></details>

Well I think this a bit more discussion. We may add ci flag, or write a separate script for CI, although reusing demo scripts makes it at least partly tested.

---

*[build/portforward.sh, line 17 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/359#-KoBXq-3Koos0dTnScjE:-KoTyB3Y1TvHsnag-QF0:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/f2c7183f406f5e3de9f97d2b27162806f7a1c364/build/portforward.sh#L17)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Even if output value it's not used now - this should end with `exit 1` or any other  than `0` value.
</blockquote></details>

Done.

---

*[pkg/criproxy/proxy_test.go, line 1160 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/359#-KoBYk36O2QFjs0Yj9Yl:-KoTyeIjRcOix4qyOkUP:bmcnafs) ([raw file](https://github.com/mirantis/virtlet/blob/f2c7183f406f5e3de9f97d2b27162806f7a1c364/pkg/criproxy/proxy_test.go#L1160)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

This also could be written in form like:
```
func skip() {
    t.Skip("skipping flaky test -- will only run on master branch")
}

switch {
case os.Getenv("CIRCLECI") == "true":
    if os.Getenv("CIRCLE_PULL_REQUEST") != "" || os.Getenv("CIRCLE_BRANCH") != "master" {
        skip()
    }
case osGetenv("TRAVIS") == "true":
    if os.Getenv("TRAVIS_PULL_REQUEST") != "false" || os.Getenv("TRAVIS_BRANCH") != "master" {
        skip()
    }
}
```
With that we have a single point of exit and IMO it's a bit more readable.
</blockquote></details>

Done. (although used a bit different name instead of generic 'skip')

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/359)*

**Comment by jellonek** [comment]
:lgtm:

---

Review status: 1 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[.circleci/config.yml, line 62 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/359#-Ko87nuokqODRW398qK6:-KoWYqhTob19jC8MzTOD:bb6mt9b) ([raw file](https://github.com/mirantis/virtlet/blob/36ace811bde339d6bf06b5833fde4b84b34de8c9/.circleci/config.yml#L62)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

Well I think this a bit more discussion. We may add ci flag, or write a separate script for CI, although reusing demo scripts makes it at least partly tested.
</blockquote></details>

:+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/359#-:-KoWYtt7-hqqI-XUEH2c:bnfp4nl)*

**Pull request merged**